Julieta Madrigal Flores - 744029

Jueves 19 de febrero del 2026

T05 - Motor Trend Car Road Tests

##**1.1 Realiza una regresión tomando 'mpg' como salida y eliminando la columna 'model'. Considera todos los demás factores como numéricos/ordinales.**


In [2]:
import pandas as pd
import numpy as np

#-EDA

In [3]:
data = pd.read_excel('Motor Trend Car Road Tests.xlsx')
data.head()

model   mpg  cyl   disp   hp  drat     wt   qsec  vs  am  gear  \
0          Mazda RX4  21.0    6  160.0  110  3.90  2.620  16.46   0   1     4   
1      Mazda RX4 Wag  21.0    6  160.0  110  3.90  2.875  17.02   0   1     4   
2         Datsun 710  22.8    4  108.0   93  3.85  2.320  18.61   1   1     4   
3     Hornet 4 Drive  21.4    6  258.0  110  3.08  3.215  19.44   1   0     3   
4  Hornet Sportabout  18.7    8  360.0  175  3.15  3.440  17.02   0   0     3   

   carb  
0     4  
1     4  
2     1  
3     1  
4     2

In [4]:
data.describe()

mpg        cyl        disp          hp       drat         wt  \
count  32.000000  32.000000   32.000000   32.000000  32.000000  32.000000   
mean   20.090625   6.187500  230.721875  146.687500   3.596563   3.217250   
std     6.026948   1.785922  123.938694   68.562868   0.534679   0.978457   
min    10.400000   4.000000   71.100000   52.000000   2.760000   1.513000   
25%    15.425000   4.000000  120.825000   96.500000   3.080000   2.581250   
50%    19.200000   6.000000  196.300000  123.000000   3.695000   3.325000   
75%    22.800000   8.000000  326.000000  180.000000   3.920000   3.610000   
max    33.900000   8.000000  472.000000  335.000000   4.930000   5.424000   

            qsec         vs         am       gear     carb  
count  32.000000  32.000000  32.000000  32.000000  32.0000  
mean   17.848750   0.437500   0.406250   3.687500   2.8125  
std     1.786943   0.504016   0.498991   0.737804   1.6152  
min    14.500000   0.000000   0.000000   3.000000   1.0000  
25%    16.892500   0.000000   0.000000   3.000000   2.0000  
50%    17.710000   0.000000   0.000000   4.000000   2.0000  
75%    18.900000   1.000000   1.000000   4.000000   4.0000  
max    22.900000   1.000000   1.000000   5.000000   8.0000

#- Definimos variables X y Y

In [5]:
y = data['mpg']
y

0     21.0
1     21.0
2     22.8
3     21.4
4     18.7
5     18.1
6     14.3
7     24.4
8     22.8
9     19.2
10    17.8
11    16.4
12    17.3
13    15.2
14    10.4
15    10.4
16    14.7
17    32.4
18    30.4
19    33.9
20    21.5
21    15.5
22    15.2
23    13.3
24    19.2
25    27.3
26    26.0
27    30.4
28    15.8
29    19.7
30    15.0
31    21.4
Name: mpg, dtype: float64

In [6]:
# hago mi matriz de X sin los boleanos para poder estandarizar
X = data.drop(['mpg', 'model', 'am', 'vs'], axis=1)
X

cyl   disp   hp  drat     wt   qsec  gear  carb
0     6  160.0  110  3.90  2.620  16.46     4     4
1     6  160.0  110  3.90  2.875  17.02     4     4
2     4  108.0   93  3.85  2.320  18.61     4     1
3     6  258.0  110  3.08  3.215  19.44     3     1
4     8  360.0  175  3.15  3.440  17.02     3     2
5     6  225.0  105  2.76  3.460  20.22     3     1
6     8  360.0  245  3.21  3.570  15.84     3     4
7     4  146.7   62  3.69  3.190  20.00     4     2
8     4  140.8   95  3.92  3.150  22.90     4     2
9     6  167.6  123  3.92  3.440  18.30     4     4
10    6  167.6  123  3.92  3.440  18.90     4     4
11    8  275.8  180  3.07  4.070  17.40     3     3
12    8  275.8  180  3.07  3.730  17.60     3     3
13    8  275.8  180  3.07  3.780  18.00     3     3
14    8  472.0  205  2.93  5.250  17.98     3     4
15    8  460.0  215  3.00  5.424  17.82     3     4
16    8  440.0  230  3.23  5.345  17.42     3     4
17    4   78.7   66  4.08  2.200  19.47     4     1
18    4   75.7   52  4.93  1.615  18.52     4     2
19    4   71.1   65  4.22  1.835  19.90     4     1
20    4  120.1   97  3.70  2.465  20.01     3     1
21    8  318.0  150  2.76  3.520  16.87     3     2
22    8  304.0  150  3.15  3.435  17.30     3     2
23    8  350.0  245  3.73  3.840  15.41     3     4
24    8  400.0  175  3.08  3.845  17.05     3     2
25    4   79.0   66  4.08  1.935  18.90     4     1
26    4  120.3   91  4.43  2.140  16.70     5     2
27    4   95.1  113  3.77  1.513  16.90     5     2
28    8  351.0  264  4.22  3.170  14.50     5     4
29    6  145.0  175  3.62  2.770  15.50     5     6
30    8  301.0  335  3.54  3.570  14.60     5     8
31    4  121.0  109  4.11  2.780  18.60     4     2

#-Estandarización

In [7]:
# Estandarizamos
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
scaler.fit(X)
X_estandarizar= scaler.transform(X)


In [8]:
# Verficamos que Media = 0 y Desviacion estandar = 1
X_estandarizar = pd.DataFrame(X_estandarizar)
X_estandarizar.describe()

0             1          2             3             4  \
count  3.200000e+01  3.200000e+01  32.000000  3.200000e+01  3.200000e+01   
mean   2.775558e-17  1.665335e-16   0.000000 -3.053113e-16 -4.718448e-16   
std    1.016001e+00  1.016001e+00   1.016001  1.016001e+00  1.016001e+00   
min   -1.244457e+00 -1.308518e+00  -1.403130 -1.589643e+00 -1.769642e+00   
25%   -1.244457e+00 -9.008917e-01  -0.743705 -9.815764e-01 -6.604034e-01   
50%   -1.066677e-01 -2.821771e-01  -0.351014  1.870518e-01  1.118844e-01   
75%    1.031121e+00  7.810529e-01   0.493642  6.145986e-01  4.078199e-01   
max    1.031121e+00  1.977904e+00   2.790515  2.533809e+00  2.291423e+00   

                  5             6             7  
count  3.200000e+01  3.200000e+01  3.200000e+01  
mean  -1.471046e-15 -2.775558e-17 -2.775558e-17  
std    1.016001e+00  1.016001e+00  1.016001e+00  
min   -1.903996e+00 -9.467293e-01 -1.140108e+00  
25%   -5.436944e-01 -9.467293e-01 -5.110827e-01  
50%   -7.888899e-02  4.303315e-01 -5.110827e-01  
75%    5.977084e-01  4.303315e-01  7.469671e-01  
max    2.871986e+00  1.807392e+00  3.263067e+00

In [9]:
# Lo hacemos en formato numpy para que se pueda concatenar con el data estadarizado
data[['vs', 'am']].values

array([[0, 1],
       [0, 1],
       [1, 1],
       [1, 0],
       [0, 0],
       [1, 0],
       [0, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [1, 1],
       [1, 1],
       [1, 1],
       [1, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [1, 1],
       [0, 1],
       [1, 1],
       [0, 1],
       [0, 1],
       [0, 1],
       [1, 1]])

In [10]:
# Hacemos X matriz y le agregamos las columnas boleanas
X = np.hstack([X_estandarizar, data[['vs', 'am']].values])
X

array([[-1.06667720e-01, -5.79750316e-01, -5.43654869e-01,
         5.76594481e-01, -6.20166581e-01, -7.89600577e-01,
         4.30331483e-01,  7.46967077e-01,  0.00000000e+00,
         1.00000000e+00],
       [-1.06667720e-01, -5.79750316e-01, -5.43654869e-01,
         5.76594481e-01, -3.55382189e-01, -4.71201785e-01,
         4.30331483e-01,  7.46967077e-01,  0.00000000e+00,
         1.00000000e+00],
       [-1.24445674e+00, -1.00602601e+00, -7.95569902e-01,
         4.81584062e-01, -9.31677630e-01,  4.32823359e-01,
         4.30331483e-01, -1.14010764e+00,  1.00000000e+00,
         1.00000000e+00],
       [-1.06667720e-01,  2.23615417e-01, -5.43654869e-01,
        -9.81576392e-01, -2.33633287e-03,  9.04735855e-01,
        -9.46729262e-01, -1.14010764e+00,  1.00000000e+00,
         0.00000000e+00],
       [ 1.03112130e+00,  1.05977159e+00,  4.19549669e-01,
        -8.48561806e-01,  2.31296954e-01, -4.71201785e-01,
        -9.46729262e-01, -5.11082737e-01,  0.00000000e+00,
         0.

#-Regresión sin penalización

In [11]:
import statsmodels.api as sm
X_con_unos= sm.add_constant(X)
model = sm.OLS(y, X_con_unos)
results = model.fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    mpg   R-squared:                       0.869
Model:                            OLS   Adj. R-squared:                  0.807
Method:                 Least Squares   F-statistic:                     13.93
Date:                Thu, 19 Feb 2026   Prob (F-statistic):           3.79e-07
Time:                        01:50:57   Log-Likelihood:                -69.855
No. Observations:                  32   AIC:                             161.7
Df Residuals:                      21   BIC:                             177.8
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         18.9278      1.445     13.099      0.000      15.923      21.933
x1            -0.1959      1.837     -0.107      0.916      -4.016       3.624
x2             1.6267      2.178      0.747      0.463      -2.903       6.157
x3            -1.4497      1.469     -0.987      0.335      -4.505       1.605
x4             0.4142      0.861      0.481      0.635      -1.376       2.204
x5            -3.5780      1.824     -1.961      0.063      -7.372       0.216
x6             1.4440      1.285      1.123      0.274      -1.229       4.117
x7             0.4760      1.084      0.439      0.665      -1.779       2.731
x8            -0.3170      1.318     -0.241      0.812      -3.057       2.423
x9             0.3178      2.105      0.151      0.881      -4.059       4.694
x10            2.5202      2.057      1.225      0.234      -1.757       6.797
==============================================================================
Omnibus:                        1.907   Durbin-Watson:                   1.861
Prob(Omnibus):                  0.385   Jarque-Bera (JB):                1.747
Skew:                           0.521   Prob(JB):                        0.418
Kurtosis:                       2.526   Cond. No.                         14.6
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

#- Calcula el R2 e interpreta los signos de los betas

In [12]:
r2 = results.rsquared
r2

np.float64(0.8690157644777647)

A partir de los resultados obtenidos, las variables con efecto negativo son x1, x3, x5 y x8, esto significa que un incremento en estas varibles se vería reflejado en una disminución de millas por galón. El coeficiente que más llama la atención por ser el más negativo del modelo es el x5 = peso en toneladas con -3.5780, esto significa que entre más pesado sea el coche, el rendimiento de millas por galón disminuye; además, tiene un p-value de 0.063, muy cerca de ser considerado estadísticamente significativo. Por otro lado, las variables x2, x4, x6, x7, x9 y x10 presentan coeficientes positivos, destacando x10 = carb con el mayor coeficiente = 2.5202, sugieriendo que un incremento en esta variable, mejoraría el rendimiento del combustible.

Sin embargo, ninguno de los coeficientes muestran significancia estadística, ya que todos los p-values están por encima del 0.05; esto significa que a pesar de que el modelo tiene una R² alta (0.807), no hay suficiente evidencia estadística para afirmar que las variables individuales tengan un efecto real sobre la variable dependiente. Estos resultados sugieren que exite colinealidad entre las variables predictoras o que la habilidad predictiva del modelo surge a partir de la combinación de las varibles más que por el efecto individual de las variables.

#- Realiza un train-test-split donde se use el 40% de los datos para entrenar.

In [13]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.6, random_state=42)

In [14]:
# NO tiene la fila de unos
X_train.shape

(12, 10)

#-Calcula el R2 de entrenamiento y de prueba.


*Entrenamiento*

In [15]:
X_train_unos= sm.add_constant(X_train)
model_train = sm.OLS(y_train, X_train_unos)
results_train = model_train.fit()
results_train.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    mpg   R-squared:                       0.998
Model:                            OLS   Adj. R-squared:                  0.980
Method:                 Least Squares   F-statistic:                     55.84
Date:                Thu, 19 Feb 2026   Prob (F-statistic):              0.104
Time:                        01:50:57   Log-Likelihood:                -2.6714
No. Observations:                  12   AIC:                             27.34
Df Residuals:                       1   BIC:                             32.68
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         32.2709      4.891      6.598      0.096     -29.874      94.415
x1            -5.3770      2.083     -2.582      0.235     -31.840      21.086
x2            -1.5453      3.482     -0.444      0.734     -45.793      42.702
x3            -2.1789      1.084     -2.011      0.294     -15.948      11.590
x4            -3.1739      0.765     -4.150      0.151     -12.892       6.544
x5            -9.8156      1.572     -6.244      0.101     -29.790      10.159
x6             9.7366      1.749      5.566      0.113     -12.492      31.965
x7             2.4961      0.711      3.510      0.177      -6.539      11.532
x8             9.4577      2.131      4.438      0.141     -17.618      36.534
x9           -26.3463      8.698     -3.029      0.203    -136.861      84.168
x10            4.4880      1.547      2.901      0.211     -15.171      24.147
==============================================================================
Omnibus:                        4.750   Durbin-Watson:                   1.590
Prob(Omnibus):                  0.093   Jarque-Bera (JB):                1.604
Skew:                          -0.411   Prob(JB):                        0.449
Kurtosis:                       4.591   Cond. No.                         87.7
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [16]:
r2_train = results_train.rsquared
r2_train

np.float64(0.9982122857062323)

*Prueba*

In [17]:
X_test_unos= sm.add_constant(X_test)
model_test = sm.OLS(y_test, X_test_unos)
results_test = model_test.fit()
results_test.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    mpg   R-squared:                       0.907
Model:                            OLS   Adj. R-squared:                  0.805
Method:                 Least Squares   F-statistic:                     8.824
Date:                Thu, 19 Feb 2026   Prob (F-statistic):            0.00155
Time:                        01:50:57   Log-Likelihood:                -36.758
No. Observations:                  20   AIC:                             95.52
Df Residuals:                       9   BIC:                             106.5
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         16.8726      1.919      8.791      0.000      12.531      21.214
x1             7.5567      4.296      1.759      0.112      -2.161      17.275
x2             0.4127      2.357      0.175      0.865      -4.919       5.745
x3            -1.8000      1.752     -1.027      0.331      -5.763       2.163
x4             2.6676      1.422      1.876      0.093      -0.549       5.884
x5            -0.8495      2.172     -0.391      0.705      -5.762       4.063
x6             2.1756      1.472      1.478      0.174      -1.155       5.506
x7             3.8316      2.239      1.711      0.121      -1.233       8.897
x8            -3.6847      1.956     -1.884      0.092      -8.109       0.740
x9             2.7997      2.490      1.125      0.290      -2.832       8.432
x10            5.4073      2.880      1.878      0.093      -1.107      11.922
==============================================================================
Omnibus:                        3.137   Durbin-Watson:                   1.921
Prob(Omnibus):                  0.208   Jarque-Bera (JB):                1.655
Skew:                           0.683   Prob(JB):                        0.437
Kurtosis:                       3.348   Cond. No.                         24.7
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [18]:
r2_test = results_test.rsquared
r2_test


np.float64(0.90744478048878)

#- Añade regularización L2 (Ridge) con un hiperparámetro lambda decidido por ti. Cambia este valor y compara con varios distintos los R2 de entrenamiento y de prueba

In [19]:
from sklearn.linear_model import Ridge

# *Entrenamiento*

Para $\alpha$ = 0.01

In [20]:
# Se usa la matriz X pero SIN la fila de unos
ridge_0_01_train = Ridge(alpha=0.01)
ridge_0_01_train.fit(X_train, y_train)
r2_ridge_0_01_train = ridge_0_01_train.score(X_train, y_train)

r2_ridge_0_01_train, ridge_0_01_train.intercept_, ridge_0_01_train.coef_

(0.9949018252589862,
 np.float64(25.901982095043422),
 array([ -3.06819189,   2.08297818,  -2.51366233,  -2.59487169,
         -9.92884751,   8.01472732,   2.10707958,   6.8470744 ,
        -14.83061252,   4.31825942]))

Para $\alpha$ = 0.1

In [21]:
ridge_0_1_train = Ridge(alpha=0.1)
ridge_0_1_train.fit(X_train, y_train)
r2_ridge_0_1_train = ridge_0_1_train.score(X_train, y_train)

r2_ridge_0_1_train, ridge_0_1_train.intercept_, ridge_0_1_train.coef_

(0.9765399621695591,
 np.float64(19.963659515632568),
 array([-0.98625155,  3.43634181, -2.53395664, -1.45510667, -7.66236493,
         4.59375161,  1.54425071,  2.89901529, -3.50220681,  3.474635  ]))

Para $\alpha$ = 1

In [22]:
ridge_1_train = Ridge(alpha=1)
ridge_1_train.fit(X_train, y_train)
r2_ridge_1_train = ridge_1_train.score(X_train, y_train)

r2_ridge_1_train, ridge_1_train.intercept_, ridge_1_train.coef_

(0.9236921871877979,
 np.float64(19.074355905988558),
 array([-0.82575324,  0.09981027, -1.65285003, -0.03768674, -3.14297446,
         0.95341358,  1.09257555, -0.42550855, -0.37909362,  1.59469938]))

Para $\alpha$ = 10

In [23]:
ridge_10_train = Ridge(alpha=10)
ridge_10_train.fit(X_train, y_train)
r2_ridge_10_train = ridge_10_train.score(X_train, y_train)

r2_ridge_10_train, ridge_10_train.intercept_, ridge_10_train.coef_


(0.870554747829006,
 np.float64(19.36271563144631),
 array([-0.92970427, -0.93847326, -0.92178249,  0.62845726, -1.47010791,
         0.30392522,  0.7159976 , -0.69917549,  0.27077913,  0.5769169 ]))

# *Prueba*

Para $\alpha$ = 0.01

In [24]:
# Se usa la matriz X pero SIN la fila de unos
# SE USA LA DE TRAIN correspondiente = con el mismo alfa
r2_ridge_0_01_test = ridge_0_01_train.score(X_test, y_test)

r2_ridge_0_01_test

-2.2266088153676318

Para $\alpha$ = 0.1

In [25]:
r2_ridge_0_1_test = ridge_0_1_train.score(X_test, y_test)

r2_ridge_0_1_test

0.24983503207635982

Para $\alpha$ = 1

In [26]:
r2_ridge_1_test = ridge_1_train.score(X_test, y_test)

r2_ridge_1_test

0.6703248275317881

Para $\alpha$ = 10

In [27]:
r2_ridge_10_test = ridge_10_train.score(X_test, y_test)

r2_ridge_10_test

0.7966360056810016

#*Tabla comparativa*

In [28]:
tabla = pd.DataFrame({
    'α': ['0.01', '0.1', '1', '10'],
    'R² Train': [r2_ridge_0_01_train, r2_ridge_0_1_train, r2_ridge_1_train, r2_ridge_10_train],
    'R² Test': [r2_ridge_0_01_test, r2_ridge_0_1_test, r2_ridge_1_test, r2_ridge_10_test]
})

tabla

α  R² Train   R² Test
0  0.01  0.994902 -2.226609
1   0.1  0.976540  0.249835
2     1  0.923692  0.670325
3    10  0.870555  0.796636

##**1.2 Repite el ejercicio anterior usando 'qsec' como salida.**

#- Definimos variables X y Y

In [29]:
y2 = data['qsec']
y2

0     16.46
1     17.02
2     18.61
3     19.44
4     17.02
5     20.22
6     15.84
7     20.00
8     22.90
9     18.30
10    18.90
11    17.40
12    17.60
13    18.00
14    17.98
15    17.82
16    17.42
17    19.47
18    18.52
19    19.90
20    20.01
21    16.87
22    17.30
23    15.41
24    17.05
25    18.90
26    16.70
27    16.90
28    14.50
29    15.50
30    14.60
31    18.60
Name: qsec, dtype: float64

In [30]:
# sin boleeanos para estandarizar
X2 = data.drop(['qsec', 'model','vs', 'am'], axis=1)

#-Estandarizamos

In [31]:
scaler = StandardScaler()
scaler.fit(X2)
X2_estandarizar= scaler.transform(X2)


In [32]:
# Verficamos que Media = 0 y Desviacion estandar = 1
X2_estandarizar = pd.DataFrame(X2_estandarizar, columns=X2.columns)
X2_estandarizar.describe()

mpg           cyl          disp         hp          drat  \
count  3.200000e+01  3.200000e+01  3.200000e+01  32.000000  3.200000e+01   
mean  -4.996004e-16  2.775558e-17  1.665335e-16   0.000000 -3.053113e-16   
std    1.016001e+00  1.016001e+00  1.016001e+00   1.016001  1.016001e+00   
min   -1.633610e+00 -1.244457e+00 -1.308518e+00  -1.403130 -1.589643e+00   
25%   -7.865141e-01 -1.244457e+00 -9.008917e-01  -0.743705 -9.815764e-01   
50%   -1.501383e-01 -1.066677e-01 -2.821771e-01  -0.351014  1.870518e-01   
75%    4.567366e-01  1.031121e+00  7.810529e-01   0.493642  6.145986e-01   
max    2.327934e+00  1.031121e+00  1.977904e+00   2.790515  2.533809e+00   

                 wt          gear          carb  
count  3.200000e+01  3.200000e+01  3.200000e+01  
mean  -4.718448e-16 -2.775558e-17 -2.775558e-17  
std    1.016001e+00  1.016001e+00  1.016001e+00  
min   -1.769642e+00 -9.467293e-01 -1.140108e+00  
25%   -6.604034e-01 -9.467293e-01 -5.110827e-01  
50%    1.118844e-01  4.303315e-01 -5.110827e-01  
75%    4.078199e-01  4.303315e-01  7.469671e-01  
max    2.291423e+00  1.807392e+00  3.263067e+00

In [33]:
# Lo hacemos en formato numpy para que se pueda concatenar con el data estadarizado
data[['vs', 'am']].values
# Hacemos X matriz y le agregamos las columnas boleanas
X2 = np.hstack([X2_estandarizar, data[['vs', 'am']].values])


#-Regresión

In [34]:
X2_con_unos= sm.add_constant(X2)
model2 = sm.OLS(y2, X2_con_unos)
results2 = model2.fit()
results2.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   qsec   R-squared:                       0.875
Model:                            OLS   Adj. R-squared:                  0.815
Method:                 Least Squares   F-statistic:                     14.66
Date:                Thu, 19 Feb 2026   Prob (F-statistic):           2.44e-07
Time:                        01:50:58   Log-Likelihood:                -30.242
No. Observations:                  32   AIC:                             82.48
Df Residuals:                      21   BIC:                             98.61
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         17.7905      0.425     41.834      0.000      16.906      18.675
x1             0.4096      0.365      1.123      0.274      -0.349       1.168
x2            -0.6375      0.514     -1.239      0.229      -1.707       0.432
x3            -0.9150      0.608     -1.505      0.147      -2.180       0.350
x4            -0.1054      0.435     -0.242      0.811      -1.010       0.800
x5            -0.0690      0.250     -0.275      0.786      -0.590       0.452
x6             1.4410      0.482      2.990      0.007       0.439       2.443
x7            -0.1462      0.314     -0.465      0.647      -0.800       0.507
x8            -0.4350      0.371     -1.174      0.254      -1.206       0.336
x9             0.9700      0.573      1.694      0.105      -0.221       2.161
x10           -0.9012      0.585     -1.540      0.139      -2.118       0.316
==============================================================================
Omnibus:                       21.069   Durbin-Watson:                   2.573
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               38.291
Skew:                           1.470   Prob(JB):                     4.84e-09
Kurtosis:                       7.481   Cond. No.                         15.3
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

#- Calcula el R2 e interpreta los signos de los betas

In [35]:
X2_r2 = results2.rsquared
X2_r2

np.float64(0.8746925777093998)


A partir de los resultados obtenidos se puede observar que solo la variable x6 y la constante son estadísticamente significativos al 5% modelo, esto sugiere que cuando todas las variables predictoras son cero, el valor esperado es de 17.79; mientras que por cada unidad adicional en x6, qsec aumenta 1.44. Por otro lado, el resto de las variables predictoras no cumplen con la significancia estdaística, lo cual sugiere que existe una posible multicolinealidad que podría estar afectando el efecto individual de cada variable.

#- Realiza un train-test-split donde se use el 40% de los datos para entrenar.

In [68]:
X2_train, X2_test, y2_train, y2_test = train_test_split(X2, y2, test_size=0.6, random_state=500)

#-Calcula el R2 de entrenamiento y de prueba.

*Entrenamiento*

In [37]:
X2_train_unos= sm.add_constant(X2_train)
model_train2 = sm.OLS(y2_train, X2_train_unos)
results_train2 = model_train2.fit()
results_train2.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   qsec   R-squared:                       0.902
Model:                            OLS   Adj. R-squared:                 -0.080
Method:                 Least Squares   F-statistic:                    0.9181
Date:                Thu, 19 Feb 2026   Prob (F-statistic):              0.679
Time:                        01:50:58   Log-Likelihood:                -12.028
No. Observations:                  12   AIC:                             46.06
Df Residuals:                       1   BIC:                             51.39
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         17.2663      2.818      6.127      0.103     -18.541      53.073
x1            -1.5437      4.139     -0.373      0.773     -54.130      51.042
x2            -2.6547      6.815     -0.390      0.764     -89.247      83.938
x3            -1.2408      6.467     -0.192      0.879     -83.413      80.931
x4             1.2912      3.774      0.342      0.790     -46.666      49.248
x5            -2.0747      6.725     -0.309      0.809     -87.519      83.370
x6             0.2565     10.904      0.024      0.985    -138.293     138.806
x7            -0.0490      2.052     -0.024      0.985     -26.122      26.024
x8            -0.9978      5.327     -0.187      0.882     -68.689      66.694
x9             2.3706      5.267      0.450      0.731     -64.557      69.298
x10            1.1082      4.213      0.263      0.836     -52.422      54.638
==============================================================================
Omnibus:                        3.447   Durbin-Watson:                   2.860
Prob(Omnibus):                  0.178   Jarque-Bera (JB):                1.190
Skew:                           0.723   Prob(JB):                        0.552
Kurtosis:                       3.539   Cond. No.                         57.3
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [38]:
r2_train2 = results_train2.rsquared
r2_train2

np.float64(0.9017789050313821)

*Prueba*

In [41]:
X2_test_unos= sm.add_constant(X2_test)
model_test2 = sm.OLS(y2_test, X2_test_unos)
results_test2 = model_test2.fit()
results_test2.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   qsec   R-squared:                       0.966
Model:                            OLS   Adj. R-squared:                  0.929
Method:                 Least Squares   F-statistic:                     25.95
Date:                Thu, 19 Feb 2026   Prob (F-statistic):           1.96e-05
Time:                        01:53:16   Log-Likelihood:                -2.7190
No. Observations:                  20   AIC:                             27.44
Df Residuals:                       9   BIC:                             38.39
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         16.4829      0.524     31.457      0.000      15.298      17.668
x1             0.6801      0.289      2.355      0.043       0.027       1.334
x2            -0.1505      0.445     -0.338      0.743      -1.156       0.855
x3            -0.8060      0.426     -1.894      0.091      -1.769       0.157
x4            -0.3489      0.316     -1.105      0.298      -1.063       0.366
x5            -0.3128      0.245     -1.278      0.233      -0.867       0.241
x6             1.0582      0.305      3.471      0.007       0.369       1.748
x7            -1.1011      0.388     -2.835      0.020      -1.980      -0.222
x8             0.3782      0.408      0.926      0.378      -0.545       1.302
x9             2.1814      0.570      3.827      0.004       0.892       3.471
x10            0.3429      0.600      0.572      0.581      -1.013       1.699
==============================================================================
Omnibus:                        1.896   Durbin-Watson:                   2.412
Prob(Omnibus):                  0.388   Jarque-Bera (JB):                1.019
Skew:                           0.091   Prob(JB):                        0.601
Kurtosis:                       1.909   Cond. No.                         26.8
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [42]:
r2_test2 = results_test2.rsquared
r2_test2

np.float64(0.9664845471168656)

#- Añade regularización L2 con un hiperparámetro lambda decidido por ti. Cambia este valor y compara con varios distintos los R2 de entrenamiento y de prueba

# *Entrenamiento*

Para $\alpha$ = 0.01

In [43]:
# Se usa la matriz X pero SIN la fila de unos
ridge2_0_01_train = Ridge(alpha=0.01)
ridge2_0_01_train.fit(X2_train, y2_train)
r2_ridge2_0_01_train = ridge2_0_01_train.score(X2_train, y2_train)

r2_ridge2_0_01_train, ridge2_0_01_train.intercept_, ridge2_0_01_train.coef_

(0.9014950022023981,
 np.float64(17.292692043623372),
 array([-1.39180842, -2.36285854, -1.40097185,  1.16156807, -1.79846462,
         0.64873211, -0.01542058, -1.12996735,  2.16091391,  0.97012066]))

Para $\alpha$ = 0.1

In [44]:
ridge2_0_1_train = Ridge(alpha=0.1)
ridge2_0_1_train.fit(X2_train, y2_train)
r2_ridge2_0_1_train = ridge2_0_1_train.score(X2_train, y2_train)

r2_ridge2_0_1_train, ridge2_0_1_train.intercept_, ridge2_0_1_train.coef_

(0.8940527102406989,
 np.float64(17.495920019238454),
 array([-0.89532373, -1.75088591, -1.28296748,  0.65349254, -1.19325963,
         1.04551384,  0.02575641, -1.06667785,  1.66928767,  0.32668644]))

Para $\alpha$ = 1

In [45]:
ridge2_1_train = Ridge(alpha=1)
ridge2_1_train.fit(X2_train, y2_train)
r2_ridge2_1_train = ridge2_1_train.score(X2_train, y2_train)

r2_ridge2_1_train, ridge2_1_train.intercept_, ridge2_1_train.coef_

(0.8421396903790743,
 np.float64(17.83881338066966),
 array([-0.19068441, -1.06346205, -0.50478946, -0.26001348, -0.51103865,
         0.73615982, -0.09695033, -0.49364875,  1.02602314, -0.50798125]))

Para $\alpha$ = 10

In [46]:
ridge2_10_train = Ridge(alpha=10)
ridge2_10_train.fit(X2_train, y2_train)
r2_ridge2_10_train = ridge2_10_train.score(X2_train, y2_train)

r2_ridge2_10_train, ridge2_10_train.intercept_, ridge2_10_train.coef_

(0.6547701818809976,
 np.float64(17.959315742055523),
 array([ 0.11712561, -0.48063392, -0.24668667, -0.41526824, -0.22222071,
         0.22522283, -0.18609249, -0.21662843,  0.38431804, -0.28734939]))

# *Prueba*

Para $\alpha$ = 0.01

In [47]:
# Se usa la matriz X pero SIN la fila de unos
# SE USA EL DE TRAIN CORRESPONDIENTE
# Eliminamos las betas porque se usan las de train
r2_ridge2_0_01_test = ridge2_0_01_train.score(X2_test, y2_test)

r2_ridge2_0_01_test

-1.1002529194403734

Para $\alpha$ = 0.1

In [48]:
r2_ridge2_0_1_test = ridge2_0_1_train.score(X2_test, y2_test)

r2_ridge2_0_1_test

-0.010520778976858214

Para $\alpha$ = 1

In [49]:
r2_ridge2_1_test = ridge2_1_train.score(X2_test, y2_test)

r2_ridge2_1_test

Para $\alpha$ = 10

In [50]:
r2_ridge2_10_test = ridge2_10_train.score(X2_test, y2_test)

r2_ridge2_10_test

0.7633335506579821

#*Tabla comparativa*

In [51]:
tabla2 = pd.DataFrame({
    'α': ['0.01', '0.1', '1', '10'],
    'R² Train': [r2_ridge2_0_01_train, r2_ridge2_0_1_train, r2_ridge2_1_train, r2_ridge2_10_train],
    'R² Test': [r2_ridge2_0_01_test, r2_ridge2_0_1_test, r2_ridge2_1_test, r2_ridge2_10_test]
})

tabla2

α  R² Train   R² Test
0  0.01  0.901495 -1.100253
1   0.1  0.894053 -0.010521
2     1  0.842140  0.762057
3    10  0.654770  0.763334

## **2.1 Realiza una regresión tomando 'mpg' como salida y eliminando la columna 'model'.**

#- Definir X y Y

In [52]:
ydummies = data['mpg']
Xdummies = data.drop(['mpg', 'model'], axis=1)


#-Crea columnas dummies para los factores 'cyl', 'gear' y 'carb'.

In [53]:
Xdummies = pd.get_dummies(Xdummies, columns=['cyl', 'gear', 'carb'], drop_first=True)

In [54]:
# Cuando creas las dummies con pd.get_dummies(), conviertes variables categóricas en múltiples columnas binarias de True / False
Xdummies.shape

(32, 16)

In [55]:
# Solo vamos a guardar las dummies
Xdummies_only = Xdummies.select_dtypes(include='bool')

#-Estandarizar

In [56]:
# sacamos las boleanas para estandarizar
Xdummies_sin_boleanas = data.drop(['mpg', 'vs', 'am','model' ], axis=1)

In [57]:
scaler = StandardScaler()
scaler.fit(Xdummies_sin_boleanas)
Xdummies_estandarizar= scaler.transform(Xdummies_sin_boleanas)


In [58]:
# Verficamos que Media = 0 y Desviacion estandar = 1
Xdummies_estandarizar = pd.DataFrame(Xdummies_estandarizar)
Xdummies_estandarizar.describe()

0             1          2             3             4  \
count  3.200000e+01  3.200000e+01  32.000000  3.200000e+01  3.200000e+01   
mean   2.775558e-17  1.665335e-16   0.000000 -3.053113e-16 -4.718448e-16   
std    1.016001e+00  1.016001e+00   1.016001  1.016001e+00  1.016001e+00   
min   -1.244457e+00 -1.308518e+00  -1.403130 -1.589643e+00 -1.769642e+00   
25%   -1.244457e+00 -9.008917e-01  -0.743705 -9.815764e-01 -6.604034e-01   
50%   -1.066677e-01 -2.821771e-01  -0.351014  1.870518e-01  1.118844e-01   
75%    1.031121e+00  7.810529e-01   0.493642  6.145986e-01  4.078199e-01   
max    1.031121e+00  1.977904e+00   2.790515  2.533809e+00  2.291423e+00   

                  5             6             7  
count  3.200000e+01  3.200000e+01  3.200000e+01  
mean  -1.471046e-15 -2.775558e-17 -2.775558e-17  
std    1.016001e+00  1.016001e+00  1.016001e+00  
min   -1.903996e+00 -9.467293e-01 -1.140108e+00  
25%   -5.436944e-01 -9.467293e-01 -5.110827e-01  
50%   -7.888899e-02  4.303315e-01 -5.110827e-01  
75%    5.977084e-01  4.303315e-01  7.469671e-01  
max    2.871986e+00  1.807392e+00  3.263067e+00

In [59]:
# Lo hacemos en formato numpy para que se pueda concatenar con el data estadarizado
data[['vs', 'am']].values

array([[0, 1],
       [0, 1],
       [1, 1],
       [1, 0],
       [0, 0],
       [1, 0],
       [0, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [1, 1],
       [1, 1],
       [1, 1],
       [1, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [1, 1],
       [0, 1],
       [1, 1],
       [0, 1],
       [0, 1],
       [0, 1],
       [1, 1]])

In [66]:
# Hacemos X matriz y le agregamos las columnas boleanas
Xdummies = np.hstack([Xdummies_estandarizar, Xdummies_only, data[['vs', 'am']].values])

#- Regresión sin penalización

In [61]:
Xdummies_con_unos= sm.add_constant(Xdummies)
modeldummies = sm.OLS(ydummies, Xdummies_con_unos)
resultsdummies = modeldummies.fit()
resultsdummies.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    mpg   R-squared:                       0.893
Model:                            OLS   Adj. R-squared:                  0.779
Method:                 Least Squares   F-statistic:                     7.830
Date:                Thu, 19 Feb 2026   Prob (F-statistic):           0.000124
Time:                        01:55:31   Log-Likelihood:                -66.608
No. Observations:                  32   AIC:                             167.2
Df Residuals:                      15   BIC:                             192.1
Df Model:                          16                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         14.8626      2.116      7.023      0.000      10.352      19.373
x1            -2.9807      2.700     -1.104      0.287      -8.736       2.775
x2             4.3362      3.891      1.114      0.283      -3.958      12.630
x3            -4.7580      2.661     -1.788      0.094     -10.429       0.913
x4             0.6225      1.307      0.476      0.641      -2.163       3.408
x5            -4.3624      2.445     -1.784      0.095      -9.574       0.849
x6             0.6470      1.645      0.393      0.700      -2.860       4.154
x7            -0.5317      1.295     -0.411      0.687      -3.291       2.228
x8             0.9570      1.829      0.523      0.608      -2.942       4.856
x9             0.7428      2.027      0.366      0.719      -3.577       5.062
x10            6.4467      1.558      4.137      0.001       3.125       9.768
x11            1.8465      2.577      0.717      0.485      -3.646       7.339
x12            3.9927      1.036      3.852      0.002       1.784       6.202
x13           -1.5813      2.021     -0.782      0.446      -5.889       2.726
x14            1.7957      3.148      0.570      0.577      -4.914       8.506
x15           -0.7145      2.297     -0.311      0.760      -5.610       4.181
x16            1.4676      2.807      0.523      0.609      -4.514       7.450
x17            3.0365      2.476      1.227      0.239      -2.240       8.313
x18            1.9309      2.871      0.672      0.512      -4.189       8.051
x19            1.2121      3.214      0.377      0.711      -5.637       8.062
==============================================================================
Omnibus:                        2.468   Durbin-Watson:                   2.105
Prob(Omnibus):                  0.291   Jarque-Bera (JB):                1.772
Skew:                           0.576   Prob(JB):                        0.412
Kurtosis:                       3.009   Cond. No.                     6.98e+16
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is 3.56e-32. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

#- Calcula el R2 e interpreta los signos de los betas.

In [62]:
r2dummies = resultsdummies.rsquared
r2dummies

np.float64(0.8930749320864843)

A partir de los resultados obtenidos se puede observar que solo x10 y x12 son significativas al 5%, lo cual indica que por cada unidad adicional, mpg aumenta 6.45 y 3.99 unidades respectivamente. Sin embargo, a pesar de tener una R² alto, el modelo tiene 16 variables predictoras y apenas 32 observaciones, lo cual es muy poco como para que el modelo logre generalizar correctamente, sugieriendo un sobreajuste.

#- Realiza un train-test-split donde se use el 40% de los datos para entrenar.

In [64]:
X_dummies_train, X_dummies_test, y_dummies_train, y_dummies_test = train_test_split(Xdummies, ydummies, test_size=0.6, random_state=42)

#- Calcula el R2 de entrenamiento y de prueba. - DUDAAAAAA

*Entrenamiento*

In [72]:
Xdummies_train_unos= sm.add_constant(X_dummies_train)
modeldummies_train = sm.OLS(y_dummies_train, Xdummies_train_unos)
results_train_dummies = modeldummies_train.fit()
results_train_dummies.summary()

/usr/local/lib/python3.12/dist-packages/statsmodels/regression/linear_model.py:1795: RuntimeWarning: divide by zero encountered in divide
  return 1 - (np.divide(self.nobs - self.k_constant, self.df_resid)
/usr/local/lib/python3.12/dist-packages/statsmodels/regression/linear_model.py:1795: RuntimeWarning: invalid value encountered in scalar multiply
  return 1 - (np.divide(self.nobs - self.k_constant, self.df_resid)
/usr/local/lib/python3.12/dist-packages/statsmodels/regression/linear_model.py:1717: RuntimeWarning: divide by zero encountered in scalar divide
  return np.dot(wresid, wresid) / self.df_resid
/usr/local/lib/python3.12/dist-packages/statsmodels/base/model.py:1527: RuntimeWarning: invalid value encountered in multiply
  cov_p = self.normalized_cov_params * scale


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    mpg   R-squared:                       1.000
Model:                            OLS   Adj. R-squared:                    nan
Method:                 Least Squares   F-statistic:                       nan
Date:                Thu, 19 Feb 2026   Prob (F-statistic):                nan
Time:                        02:19:52   Log-Likelihood:                 358.04
No. Observations:                  12   AIC:                            -692.1
Df Residuals:                       0   BIC:                            -686.3
Df Model:                          11                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          9.4509        inf          0        nan         nan         nan
x1            -0.1761        inf         -0        nan         nan         nan
x2             3.1398        inf          0        nan         nan         nan
x3            -3.4853        inf         -0        nan         nan         nan
x4            -3.1124        inf         -0        nan         nan         nan
x5           -10.2047        inf         -0        nan         nan         nan
x6             5.9972        inf          0        nan         nan         nan
x7             1.4864        inf          0        nan         nan         nan
x8             2.5370        inf          0        nan         nan         nan
x9            -5.3558        inf         -0        nan         nan         nan
x10            7.7689        inf          0        nan         nan         nan
x11            3.7344        inf          0        nan         nan         nan
x12            1.9212        inf          0        nan         nan         nan
x13            0.6879        inf          0        nan         nan         nan
x14                 0        nan        nan        nan         nan         nan
x15            6.8250        inf          0        nan         nan         nan
x16                 0        nan        nan        nan         nan         nan
x17                 0        nan        nan        nan         nan         nan
x18            1.6819        inf          0        nan         nan         nan
x19            3.0022        inf          0        nan         nan         nan
==============================================================================
Omnibus:                        0.204   Durbin-Watson:                   0.891
Prob(Omnibus):                  0.903   Jarque-Bera (JB):                0.382
Skew:                          -0.175   Prob(JB):                        0.826
Kurtosis:                       2.200   Cond. No.                         65.4
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The input rank is higher than the number of observations.
"""

In [ ]:
32*.6

In [73]:
r2dummies_train = results_train_dummies.rsquared
r2dummies_train

np.float64(1.0)

*Prueba*

In [74]:
Xdummies_test_unos= sm.add_constant(X_dummies_test)
modeldummies_test = sm.OLS(y_dummies_test, Xdummies_test_unos)
results_test_dummies = modeldummies_test.fit()
results_test_dummies.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    mpg   R-squared:                       0.975
Model:                            OLS   Adj. R-squared:                  0.883
Method:                 Least Squares   F-statistic:                     10.55
Date:                Thu, 19 Feb 2026   Prob (F-statistic):             0.0174
Time:                        02:19:58   Log-Likelihood:                -23.529
No. Observations:                  20   AIC:                             79.06
Df Residuals:                       4   BIC:                             94.99
Df Model:                          15                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         11.1811      2.741      4.080      0.015       3.572      18.790
x1             9.5239      8.118      1.173      0.306     -13.016      32.063
x2            -1.6364      4.968     -0.329      0.758     -15.429      12.156
x3            -4.3323      6.661     -0.650      0.551     -22.826      14.162
x4            11.7487      3.418      3.437      0.026       2.259      21.238
x5             5.4494      4.119      1.323      0.256      -5.986      16.884
x6             3.1199      4.214      0.740      0.500      -8.581      14.821
x7            -3.4946      7.517     -0.465      0.666     -24.364      17.375
x8            -2.4665      7.742     -0.319      0.766     -23.961      19.028
x9             7.6185      4.468      1.705      0.163      -4.787      20.024
x10            6.4907      2.161      3.004      0.040       0.492      12.490
x11           -1.4917      3.103     -0.481      0.656     -10.108       7.124
x12            3.3205      1.508      2.202      0.092      -0.866       7.507
x13            0.4970      3.240      0.153      0.885      -8.497       9.491
x14           -3.0249      4.140     -0.731      0.505     -14.520       8.470
x15           -9.7787      5.544     -1.764      0.153     -25.171       5.614
x16            4.3192      2.931      1.473      0.215      -3.819      12.458
x17            4.2339      5.989      0.707      0.519     -12.395      20.863
x18            4.3838      2.116      2.072      0.107      -1.491      10.259
x19           12.3796      3.245      3.814      0.019       3.369      21.391
==============================================================================
Omnibus:                        2.054   Durbin-Watson:                   1.842
Prob(Omnibus):                  0.358   Jarque-Bera (JB):                0.944
Skew:                           0.519   Prob(JB):                        0.624
Kurtosis:                       3.236   Cond. No.                     4.66e+18
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is 4.56e-36. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

In [75]:
r2dummies_test= results_test_dummies.rsquared
r2dummies_test

np.float64(0.9753468944149285)

## **2.2 Repite el ejercicio anterior usando 'qsec' como salida.**

#-Definir X y Y

In [76]:
ydummies2 = data['qsec']
Xdummies2 = data.drop(['qsec', 'model'], axis=1)

#-Crea columnas dummies para los factores 'cyl', 'gear' y 'carb'.

In [77]:
Xdummies2 = pd.get_dummies(Xdummies2, columns=['cyl', 'gear', 'carb'], drop_first=True)

In [78]:
# Cuando creas las dummies con pd.get_dummies(), conviertes variables categóricas en múltiples columnas binarias de True / False
Xdummies2

mpg   disp   hp  drat     wt  vs  am  cyl_6  cyl_8  gear_4  gear_5  \
0   21.0  160.0  110  3.90  2.620   0   1   True  False    True   False   
1   21.0  160.0  110  3.90  2.875   0   1   True  False    True   False   
2   22.8  108.0   93  3.85  2.320   1   1  False  False    True   False   
3   21.4  258.0  110  3.08  3.215   1   0   True  False   False   False   
4   18.7  360.0  175  3.15  3.440   0   0  False   True   False   False   
5   18.1  225.0  105  2.76  3.460   1   0   True  False   False   False   
6   14.3  360.0  245  3.21  3.570   0   0  False   True   False   False   
7   24.4  146.7   62  3.69  3.190   1   0  False  False    True   False   
8   22.8  140.8   95  3.92  3.150   1   0  False  False    True   False   
9   19.2  167.6  123  3.92  3.440   1   0   True  False    True   False   
10  17.8  167.6  123  3.92  3.440   1   0   True  False    True   False   
11  16.4  275.8  180  3.07  4.070   0   0  False   True   False   False   
12  17.3  275.8  180  3.07  3.730   0   0  False   True   False   False   
13  15.2  275.8  180  3.07  3.780   0   0  False   True   False   False   
14  10.4  472.0  205  2.93  5.250   0   0  False   True   False   False   
15  10.4  460.0  215  3.00  5.424   0   0  False   True   False   False   
16  14.7  440.0  230  3.23  5.345   0   0  False   True   False   False   
17  32.4   78.7   66  4.08  2.200   1   1  False  False    True   False   
18  30.4   75.7   52  4.93  1.615   1   1  False  False    True   False   
19  33.9   71.1   65  4.22  1.835   1   1  False  False    True   False   
20  21.5  120.1   97  3.70  2.465   1   0  False  False   False   False   
21  15.5  318.0  150  2.76  3.520   0   0  False   True   False   False   
22  15.2  304.0  150  3.15  3.435   0   0  False   True   False   False   
23  13.3  350.0  245  3.73  3.840   0   0  False   True   False   False   
24  19.2  400.0  175  3.08  3.845   0   0  False   True   False   False   
25  27.3   79.0   66  4.08  1.935   1   1  False  False    True   False   
26  26.0  120.3   91  4.43  2.140   0   1  False  False   False    True   
27  30.4   95.1  113  3.77  1.513   1   1  False  False   False    True   
28  15.8  351.0  264  4.22  3.170   0   1  False   True   False    True   
29  19.7  145.0  175  3.62  2.770   0   1   True  False   False    True   
30  15.0  301.0  335  3.54  3.570   0   1  False   True   False    True   
31  21.4  121.0  109  4.11  2.780   1   1  False  False    True   False   

    carb_2  carb_3  carb_4  carb_6  carb_8  
0    False   False    True   False   False  
1    False   False    True   False   False  
2    False   False   False   False   False  
3    False   False   False   False   False  
4     True   False   False   False   False  
5    False   False   False   False   False  
6    False   False    True   False   False  
7     True   False   False   False   False  
8     True   False   False   False   False  
9    False   False    True   False   False  
10   False   False    True   False   False  
11   False    True   False   False   False  
12   False    True   False   False   False  
13   False    True   False   False   False  
14   False   False    True   False   False  
15   False   False    True   False   False  
16   False   False    True   False   False  
17   False   False   False   False   False  
18    True   False   False   False   False  
19   False   False   False   False   False  
20   False   False   False   False   False  
21    True   False   False   False   False  
22    True   False   False   False   False  
23   False   False    True   False   False  
24    True   False   False   False   False  
25   False   False   False   False   False  
26    True   False   False   False   False  
27    True   False   False   False   False  
28   False   False    True   False   False  
29   False   False   False    True   False  
30   False   False   False   False    True  
31    True   False   False   False   False

In [79]:
# Solo vamos a guardar las dummies
Xdummies_only2 = Xdummies2.select_dtypes(include='bool')

#- Estandarizar

In [80]:
# sacamos las boleanas para estandarizar
Xdummies_sin_boleanas2 = data.drop(['qsec', 'vs', 'am','model' ], axis=1)

In [81]:
scaler = StandardScaler()
scaler.fit(Xdummies_sin_boleanas2)
Xdummies_estandarizar2= scaler.transform(Xdummies_sin_boleanas2)

In [82]:
# Verficamos que Media = 0 y Desviacion estandar = 1
Xdummies_estandarizar2 = pd.DataFrame(Xdummies_estandarizar2)
Xdummies_estandarizar2.describe()

0             1             2          3             4  \
count  3.200000e+01  3.200000e+01  3.200000e+01  32.000000  3.200000e+01   
mean  -4.996004e-16  2.775558e-17  1.665335e-16   0.000000 -3.053113e-16   
std    1.016001e+00  1.016001e+00  1.016001e+00   1.016001  1.016001e+00   
min   -1.633610e+00 -1.244457e+00 -1.308518e+00  -1.403130 -1.589643e+00   
25%   -7.865141e-01 -1.244457e+00 -9.008917e-01  -0.743705 -9.815764e-01   
50%   -1.501383e-01 -1.066677e-01 -2.821771e-01  -0.351014  1.870518e-01   
75%    4.567366e-01  1.031121e+00  7.810529e-01   0.493642  6.145986e-01   
max    2.327934e+00  1.031121e+00  1.977904e+00   2.790515  2.533809e+00   

                  5             6             7  
count  3.200000e+01  3.200000e+01  3.200000e+01  
mean  -4.718448e-16 -2.775558e-17 -2.775558e-17  
std    1.016001e+00  1.016001e+00  1.016001e+00  
min   -1.769642e+00 -9.467293e-01 -1.140108e+00  
25%   -6.604034e-01 -9.467293e-01 -5.110827e-01  
50%    1.118844e-01  4.303315e-01 -5.110827e-01  
75%    4.078199e-01  4.303315e-01  7.469671e-01  
max    2.291423e+00  1.807392e+00  3.263067e+00

In [83]:
# Lo hacemos en formato numpy para que se pueda concatenar con el data estadarizado
data[['vs', 'am']].values


array([[0, 1],
       [0, 1],
       [1, 1],
       [1, 0],
       [0, 0],
       [1, 0],
       [0, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [1, 1],
       [1, 1],
       [1, 1],
       [1, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [0, 0],
       [1, 1],
       [0, 1],
       [1, 1],
       [0, 1],
       [0, 1],
       [0, 1],
       [1, 1]])

In [84]:
# Hacemos X matriz y le agregamos las columnas boleanas
Xdummies2 = np.hstack([Xdummies_estandarizar2, Xdummies_only2, data[['vs', 'am']].values])

#-Regresión sin penalización

In [85]:
Xdummies_con_unos2= sm.add_constant(Xdummies2)
modeldummies2 = sm.OLS(ydummies2, Xdummies_con_unos2)
resultsdummies2 = modeldummies2.fit()
resultsdummies2.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   qsec   R-squared:                       0.908
Model:                            OLS   Adj. R-squared:                  0.810
Method:                 Least Squares   F-statistic:                     9.283
Date:                Thu, 19 Feb 2026   Prob (F-statistic):           4.35e-05
Time:                        02:20:31   Log-Likelihood:                -25.252
No. Observations:                  32   AIC:                             84.50
Df Residuals:                      15   BIC:                             109.4
Df Model:                          16                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         14.1491      0.551     25.681      0.000      12.975      15.323
x1             0.1646      0.418      0.393      0.700      -0.727       1.057
x2            -3.4279      0.676     -5.074      0.000      -4.868      -1.988
x3             0.4308      1.106      0.389      0.703      -1.927       2.789
x4            -0.1394      0.804     -0.173      0.865      -1.853       1.574
x5             0.0568      0.361      0.157      0.877      -0.713       0.827
x6             0.7805      0.711      1.097      0.290      -0.735       2.297
x7            -1.0551      0.358     -2.946      0.010      -1.818      -0.292
x8            -0.8769      0.499     -1.756      0.099      -1.941       0.187
x9             2.7959      0.573      4.880      0.000       1.575       4.017
x10            4.8335      0.426     11.349      0.000       3.926       5.741
x11            2.7852      0.647      4.308      0.001       1.407       4.163
x12            3.0881      0.278     11.127      0.000       2.497       3.680
x13           -0.2828      0.553     -0.512      0.616      -1.461       0.895
x14            0.8738      0.864      1.011      0.328      -0.968       2.716
x15           -0.2923      0.575     -0.508      0.619      -1.518       0.933
x16            1.1926      0.770      1.549      0.142      -0.448       2.833
x17            2.5286      0.662      3.818      0.002       1.117       3.940
x18            0.2657      0.797      0.333      0.744      -1.434       1.965
x19           -1.6943      0.771     -2.197      0.044      -3.338      -0.050
==============================================================================
Omnibus:                       12.619   Durbin-Watson:                   2.588
Prob(Omnibus):                  0.002   Jarque-Bera (JB):               16.989
Skew:                           0.924   Prob(JB):                     0.000205
Kurtosis:                       6.054   Cond. No.                     4.07e+16
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is 1.16e-31. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

#- Calcula el R2 e interpreta los signos de los betas.

In [86]:
r2dummies2 = resultsdummies2.rsquared
r2dummies2

np.float64(0.9082689440167557)

A partir de los resultados obtenidos se puede observar que el valor esperado base cuando todas las variables predictoras son cero es de 14.15, mientras que un incremento unitario de la variable x10 implica un aumento de 4.83 en qsec pero un aumento en la varible x2 significa un reducción de 3.4279 en qsec. Asimismo las variables x9, x11 y x12 son estadísticamente signficativas con un p-value menor al 0.05 y con efectos positivos; mientras que x7 y x19 también son estadísticamente significativas pero on efectos negativos.

#- Realiza un train-test-split donde se use el 40% de los datos para entrenar.

In [87]:
X2_dummies_train, X2_dummies_test, y2_dummies_train, y2_dummies_test = train_test_split(Xdummies2, ydummies2, test_size=0.6, random_state=42)

#- Calcula el R2 de entrenamiento y de prueba.

*Entrenamiento*

In [88]:
Xdummies_train_unos2= sm.add_constant(X2_dummies_train)
modeldummies_train2 = sm.OLS(y2_dummies_train, Xdummies_train_unos2)
results_train_dummies2 = modeldummies_train2.fit()
results_train_dummies2.summary()

/usr/local/lib/python3.12/dist-packages/statsmodels/regression/linear_model.py:1795: RuntimeWarning: divide by zero encountered in divide
  return 1 - (np.divide(self.nobs - self.k_constant, self.df_resid)
/usr/local/lib/python3.12/dist-packages/statsmodels/regression/linear_model.py:1795: RuntimeWarning: invalid value encountered in scalar multiply
  return 1 - (np.divide(self.nobs - self.k_constant, self.df_resid)
/usr/local/lib/python3.12/dist-packages/statsmodels/regression/linear_model.py:1717: RuntimeWarning: divide by zero encountered in scalar divide
  return np.dot(wresid, wresid) / self.df_resid
/usr/local/lib/python3.12/dist-packages/statsmodels/base/model.py:1527: RuntimeWarning: invalid value encountered in multiply
  cov_p = self.normalized_cov_params * scale


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   qsec   R-squared:                       1.000
Model:                            OLS   Adj. R-squared:                    nan
Method:                 Least Squares   F-statistic:                       nan
Date:                Thu, 19 Feb 2026   Prob (F-statistic):                nan
Time:                        02:27:37   Log-Likelihood:                 354.06
No. Observations:                  12   AIC:                            -684.1
Df Residuals:                       0   BIC:                            -678.3
Df Model:                          11                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         10.1018        inf          0        nan         nan         nan
x1             1.6183        inf          0        nan         nan         nan
x2            -0.1758        inf         -0        nan         nan         nan
x3            -1.3698        inf         -0        nan         nan         nan
x4             0.7601        inf          0        nan         nan         nan
x5             0.8446        inf          0        nan         nan         nan
x6             3.0735        inf          0        nan         nan         nan
x7            -1.5731        inf         -0        nan         nan         nan
x8            -3.4918        inf         -0        nan         nan         nan
x9             1.4641        inf          0        nan         nan         nan
x10            4.7151        inf          0        nan         nan         nan
x11            0.4247        inf          0        nan         nan         nan
x12            2.6889        inf          0        nan         nan         nan
x13            1.5320        inf          0        nan         nan         nan
x14                 0        nan        nan        nan         nan         nan
x15            3.7421        inf          0        nan         nan         nan
x16                 0        nan        nan        nan         nan         nan
x17                 0        nan        nan        nan         nan         nan
x18            5.3867        inf          0        nan         nan         nan
x19           -0.7651        inf         -0        nan         nan         nan
==============================================================================
Omnibus:                        0.175   Durbin-Watson:                   2.336
Prob(Omnibus):                  0.916   Jarque-Bera (JB):                0.066
Skew:                          -0.078   Prob(JB):                        0.968
Kurtosis:                       2.672   Cond. No.                         116.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The input rank is higher than the number of observations.
"""

In [89]:
r2dummies_train2 = results_train_dummies2.rsquared
r2dummies_train2

np.float64(1.0)

*Prueba*

In [90]:
Xdummies_test_unos2 = sm.add_constant(X2_dummies_test)
modeldummies_test2 = sm.OLS(y2_dummies_test, Xdummies_test_unos2)
results_test_dummies2 = modeldummies_test2.fit()
results_test_dummies2.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   qsec   R-squared:                       0.993
Model:                            OLS   Adj. R-squared:                  0.966
Method:                 Least Squares   F-statistic:                     36.98
Date:                Thu, 19 Feb 2026   Prob (F-statistic):            0.00158
Time:                        02:27:46   Log-Likelihood:                 9.0948
No. Observations:                  20   AIC:                             13.81
Df Residuals:                       4   BIC:                             29.74
Df Model:                          15                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         12.1305      0.264     45.869      0.000      11.396      12.865
x1             0.4030      0.544      0.740      0.500      -1.108       1.914
x2            -5.3650      0.861     -6.230      0.003      -7.756      -2.974
x3            -0.3877      0.966     -0.401      0.709      -3.070       2.294
x4            -1.5463      1.132     -1.366      0.244      -4.689       1.596
x5            -1.1938      1.189     -1.004      0.372      -4.494       2.106
x6             0.3721      0.948      0.392      0.715      -2.261       3.005
x7            -4.5659      0.868     -5.259      0.006      -6.977      -2.155
x8             3.2607      0.776      4.205      0.014       1.108       5.414
x9             0.7297      0.740      0.986      0.380      -1.326       2.785
x10            3.9114      0.427      9.161      0.001       2.726       5.097
x11           -0.3537      0.596     -0.594      0.585      -2.008       1.301
x12            2.6889      0.241     11.170      0.000       2.021       3.357
x13            3.8129      0.313     12.198      0.000       2.945       4.681
x14            2.1708      0.975      2.227      0.090      -0.535       4.877
x15            1.3446      1.568      0.857      0.440      -3.010       5.699
x16            0.1336      0.409      0.327      0.760      -1.002       1.269
x17            2.0449      0.912      2.242      0.088      -0.487       4.577
x18            3.3706      0.379      8.899      0.001       2.319       4.422
x19            1.5885      0.836      1.899      0.130      -0.733       3.910
==============================================================================
Omnibus:                        6.851   Durbin-Watson:                   2.636
Prob(Omnibus):                  0.033   Jarque-Bera (JB):                5.431
Skew:                           0.553   Prob(JB):                       0.0662
Kurtosis:                       5.301   Cond. No.                     3.39e+17
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is 9.29e-34. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

In [91]:
r2dummies_test2 = results_test_dummies2.rsquared
r2dummies_test2

np.float64(0.9928406222872872)

## **3.1 Compara los R2 de los ejercicios 1.1 & 2.1.**

In [92]:
# aquí estamos comparando la R² del modelo completo VS su R² en train VS su R² en test
# hacemos lo mismo con el modelo SIN dummies vs CON dummies

tabla_mpg = pd.DataFrame({
    'Modelo': ['Numérico ', 'Dummies '],
    'R² Total': [r2, r2dummies],
    'R² Train': [r2_train, r2dummies_train],
    'R² Test':  [r2_test, r2dummies_test]
})
tabla_mpg

Modelo  R² Total  R² Train   R² Test
0  Numérico (1.1)  0.869016  0.998212  0.907445
1   Dummies (2.1)  0.893075  1.000000  0.975347


A partir de los resultados obtenidos se puede observar que a simple vista la diferencia de R² entre Train y Test es muy pequeña y que podría significar que es un modelo confiable y que generaliza de manera correcta; sin embargo al tener una R² de práctimente 1, esto sugiere un gran sobreajuste en el modelo. Al entrenar con solo el 40% de los datos e incluir múltiples dummies que lograron fragmentar aún más la muestra, el modelo tiene tantos predictores como observaciones, logrando un "ajuste"  perfecto, haciendo a estos modelos poco confiables para un predictivo real. Es decir, es necesraio tener más observaciones o una muestra más grande para poder utilizar esta cantidad de variables predictoras en este modelo.

## **3.2 Compara los R2 de los ejercicios 1.2 & 2.2.**

In [96]:
# aquí estamos comparando la R² del modelo completo VS su R² en train VS su R² en test
# hacemos lo mismo con el modelo SIN dummies vs CON dummies

tabla_qsec = pd.DataFrame({
    'Modelo': ['Numérico', 'Dummies'],
    'R² Total': [X2_r2, r2dummies2],
    'R² Train': [r2_train2, r2dummies_train2],
    'R² Test':  [r2_test2, r2dummies_test2]
})
tabla_qsec

Modelo  R² Total  R² Train   R² Test
0  Numérico  0.874693  0.901779  0.966485
1   Dummies  0.908269  1.000000  0.992841

A partir de los resultados obtenidos se puede observar que pasó lo mismo que con el modelo anterior que utilizaba mpg como variable dependiente, donde a simple vista la diferencia de R² entre Train y Test es muy pequeña y que podría significar que es un modelo confiable y que generaliza de manera correcta; sin embargo al tener una R² de práctimente 1, esto sugiere un gran sobreajuste en el modelo. Al entrenar con solo el 40% de los datos e incluir múltiples dummies que lograron fragmentar aún más la muestra, el modelo tiene tantos predictores como observaciones, logrando un "ajuste"  perfecto, haciendo a estos modelos poco confiables para un predictivo real. Es decir, es necesraio tener más observaciones o una muestra más grande para poder utilizar esta cantidad de variables predictoras en este modelo.